core/solver/gmres.cpp

```diff
@@ -0,0 +1,230 @@
+#include "core/base/math.hpp"
+#include "core/base/name_demangling.hpp"
+#include "core/base/utils.hpp"
+#include "core/matrix/dense.hpp"
+#include "core/matrix/dense_kernels.hpp"
+#include "core/solver/gmres_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+
+
+namespace {
+
+
+template <typename ValueType>
+struct TemplatedOperation {
+    GKO_REGISTER_OPERATION(initialize_1, gmres::initialize_1<ValueType>);
+    GKO_REGISTER_OPERATION(initialize_2, gmres::initialize_2<ValueType>);
+    GKO_REGISTER_OPERATION(step_1, gmres::step_1<ValueType>);
+    GKO_REGISTER_OPERATION(step_2, gmres::step_2<ValueType>);
+};
+
+
+}  // namespace
+
+
+template <typename ValueType>
+void Gmres<ValueType>::apply_impl(const LinOp *b, LinOp *x) const
+{
+    ASSERT_IS_SQUARE_MATRIX(system_matrix_);
+
+    using Vector = matrix::Dense<ValueType>;
+
+    constexpr uint8 RelativeStoppingId{1};
+
+    auto exec = this->get_executor();
+
+    auto one_op = initialize<Vector>({one<ValueType>()}, exec);
+    auto neg_one_op = initialize<Vector>({-one<ValueType>()}, exec);
+
+    auto dense_b = as<const Vector>(b);
+    auto dense_x = as<Vector>(x);
+    auto residual = Vector::create_with_config_of(dense_b);
+    auto krylov_bases = Vector::create(
+        exec,
+        dim<2>{system_matrix_->get_size()[1],
+               (solver::default_max_iter_num + 1) * dense_b->get_size()[1]});
+    auto next_krylov_basis = Vector::create_with_config_of(dense_b);
+    auto hessenberg = Vector::create(
+        exec, dim<2>{solver::default_max_iter_num + 1,
+                     solver::default_max_iter_num * dense_b->get_size()[1]});
+    auto givens_sin = Vector::create(
+        exec, dim<2>{solver::default_max_iter_num, dense_b->get_size()[1]});
+    auto givens_cos = Vector::create(
+        exec, dim<2>{solver::default_max_iter_num, dense_b->get_size()[1]});
+    auto residual_norms = Vector::create(
+        exec, dim<2>{solver::default_max_iter_num + 1, dense_b->get_size()[1]});
+    auto residual_norm =
+        Vector::create(exec, dim<2>{1, dense_b->get_size()[1]});
+    auto b_norm = Vector::create(exec, dim<2>{1, dense_b->get_size()[1]});
+    Array<size_type> final_iter_nums(this->get_executor(),
+                                     dense_b->get_size()[1]);
+
+    bool one_changed{};
+    Array<stopping_status> stop_status(this->get_executor(),
+                                       dense_b->get_size()[1]);
+
+    // Initialization
+    exec->run(TemplatedOperation<ValueType>::make_initialize_1_operation(
+        dense_b, b_norm.get(), residual.get(), givens_sin.get(),
+        givens_cos.get(), &final_iter_nums, &stop_status,
+        solver::default_max_iter_num));
+    // b_norm = norm(b)
+    // residual = dense_b
+    // givens_sin = givens_cos = 0
+    // final_iter_nums = {0, ..., 0}
+    system_matrix_->apply(neg_one_op.get(), dense_x, one_op.get(),
+                          residual.get());
+    // residual = residual(b) - Ax
```

Comment:
I think you can write it as `residual = residual - Ax`. The `(b)` makes it a bit confusing.